Face framing

eehakkin opened the pull request:
Face framing allows automatic face framing based on the position of the subject within the field of view.

Please see https://github.com/w3c/mediacapture-extensions/pull/49#issue-1099641483 why a constrainable property is most suitable for the purpose.

yuhonglin commented:
Hi, we are also interested in supporting this in chromeOS. Is there any existing explainer/spec?

yuhonglin commented:
May I ask one question: Is there any signal/status showing the following,
1. The auto-framing is in progress.
2. The auto-framing is done.

eehakkin commented:
> Hi, we are also interested in supporting this in chromeOS. Is there any existing explainer/spec?

There is not explainer but we did create an explainer for background blur (https://github.com/riju/backgroundBlur/blob/main/explainer.md) and face framing is quite similar to that in a sense.

The spec is basically only in the CL i.e. it has not yet been approved by the W3C WebRTC WG.

alvestrand commented:
Currently waiting for the metadata format to be landed in the webcodecs metadata registry.

eehakkin commented:
> Currently waiting for the metadata format to be landed in the webcodecs metadata registry.

That would of course be nice but it is not actually directory needed for automatic face framing because the automatic human face based framing is done by the driver or by the platform which do not need webcodecs metadata.

alvestrand commented:
Is face framing intended as a "continuous follow" function?

In one trial I'm aware of, continuous follow framing was near-universally hated by the users, while a button they could hit saying "frame my face now" was popular.

The expected behavior should be clear from the functional description.

(That said, I am not at all averse to adding face framing to the spec even if it's not appropriate for many cases - this is, after all, the extensions spec, where we put stuff in order to let it be tried.)

eehakkin commented:
> Is face framing intended as a "continuous follow" function?

Yes.

> The expected behavior should be clear from the functional description.

I updated descriptions.

> (That said, I am not at all averse to adding face framing to the spec even if it's not appropriate for many cases - this is, after all, the extensions spec, where we put stuff in order to let it be tried.)

I also added a note that if a need arises, _faceFraming_ booleans could be replaced with _faceFramingMode_ `MeteringMode` `DOMString`s ("none", "manual", "single-shot", "continuous").

youennf commented on index.html:
```diff
@@ -709,6 +709,78 @@ <h2>Exposing MediaStreamTrack source background blur support</h2>
+    <div class="note">
+      If a need for automatic single-shot framing arises,
+      {{MediaTrackConstraintSet/faceFraming}} booleans could be replaced with
+      faceFramingMode {{MeteringMode}} {{DOMString}}s.
```
Do we need this note, it seems we could file an issue instead?

eehakkin replied:
Ok, I'll do that.

eehakkin commented:
I changed the example to use required constraints instead of advanced constraints.

jan-ivar reviewed (approved):
Looks ok with code sample change.

henbos commented on index.html:
```diff
@@ -709,6 +709,71 @@ <h2>Exposing MediaStreamTrack source background blur support</h2>
+      <p>When the "faceFraming" setting is set to <code>false</code> by
+         the <a>ApplyConstraints algorithm</a>, the UA will not crop to human
+         faces.
+      </p>
```
We need to clarify what happens to width and height of the track, but that can happen in a separate issue
